### ISSUE TITLE
Enhancement: [no-floating-promises] allowForKnownSafeCalls for test.only

### ISSUE THREAD

**Issue by koooge** [typescript-eslint/typescript-eslint#10740]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises

### Description

Hi there,
`@typescript-eslint/no-floating-promises` 's allowForKnownSafePromiseReturns !8404 works well for it from `node:test`. However, it doesn't help for `it.only` and `it.skip`.  They are handy during debugging so it would be great to have the ability to ignore them as well.

```js
  "allowForKnownSafePromiseReturns": [
    { "from": "package", "name": "it", "package": "node:test" }
  ]
```

### Fail

```TypeScript
import { it } from "node:test";

it.only("...", () => { /* ... */ });

it.skip("...", () => { /* ... */ });
```

### Pass

```TypeScript
import { it } from "node:test";

// eslint-disable-next-line @typescript-eslint/no-floating-promises
it.only("...", () => { /* ... */ });

// eslint-disable-next-line @typescript-eslint/no-floating-promises
it.skip("...", () => { /* ... */ });
```

**Comment by JoshuaKGoldberg** [comment]
👍 this makes sense to me as a feature request. The difficult point here is that we don't yet have a way in the shared [TypeOrValueSpecifier format](https://typescript-eslint.io/packages/type-utils/type-or-value-specifier) to specify a _property_ of something imported by a package.

Copying the current interface from https://typescript-eslint.io/packages/type-utils/type-or-value-specifier#packagespecifier:

```ts
interface PackageSpecifier {
  from: 'package';
  name: string[] | string;
  package: string;
}
```

How do we think this could be augmented to allow both `it` and `it.only`? A couple of starting ideas...


```js
{ "from": "package", "name": "it", "package": "node:test", "properties": ["skip", "only"] }
```

```js
{ "from": "package", "name": ["it", { "object": "it", "properties": ["skip", "only"] }], "package": "vitest" }
```

**Comment by kirkwaiblinger** [comment]
> How do we think this could be augmented to allow both `it` and `it.only`? A couple of starting ideas...
> 
> { "from": "package", "name": "it", "package": "node:test", "properties": ["skip", "only"] }
> { "from": "package", "name": ["it", { "object": "it", "properties": ["skip", "only"] }], "package": "vitest" }

Could we more or less use the actual JS syntax as the specifier syntax?
```json
{ "from": "package", "package": "node:test", "name": "it" }, 
{ "from": "package", "package": "node:test", "name": "it.only" },
{ "from": "package", "package": "node:test", "name": "it.skip" },
```

PS - how long until we start accepting arbitrary ESquery syntax 🙃 
```json
{ "from": "package", "package": "node:test", "selector": "Identifier[name=it]" }, 
{ "from": "package", "package": "node:test", "selector": "MemberExpression[property=only][object.type=Identifier][object.name=it]" },
{ "from": "package", "package": "node:test", "selector": "MemberExpression[property=skip][object.type=Identifier][object.name=it]" },
```

**Comment by JoshuaKGoldberg** [comment]
> Could we more or less use the actual JS syntax as the specifier syntax?

How would this extend to multiple properties? Like `(describe|it|test)` & `only|skip`?

> ESQuery syntax

😱 I never thought of using ESQuery syntax, that's a really interesting idea...

**Comment by kirkwaiblinger** [comment]
> > Could we more or less use the actual JS syntax as the specifier syntax?
> 
> How would this extend to multiple properties? Like `(describe|it|test)` & `only|skip`?
> 
> > ESQuery syntax
> 
> 😱 I never thought of using ESQuery syntax, that's a really interesting idea...

Good point.....

I mean, the other choice, which at first sounds like it wouldn't be crazy, would be regex. 
```js
/(describe|it|test)(\.(only|skip|runIf)?/
```

But this obviously fails quickly for things like 
```js
describe
   .it(
       'foo',
        () => {}
   );
```

I guess the question is would we want to be very general at the expense of being very unwieldy (direction of ESQuery), or very straightforward at the expense of being very constrained (direction of the syntaxes you've described). Trying to hit something in between seems like we'll just do a bad job at a complex problem (like the regex idea).

**Comment by Harpush** [comment]
How about:
```json
{ "from": "lib", "allowedName": "KeydownEvent" }
```
and
```json
{ "from": "lib", "name": "KeydownEvent", "allowedProperties": [ "keyCode" ] }
```
It changes quite a bit the type semantics but it is very straightforward....

Another option is just adding a note about it in docs that properties override the name (switching) and to allow both you need to specify both

**Comment by kirkwaiblinger** [comment]
@Harpush I think I'm ok with proceeding with that. The only thing I'd change is that I think we need to simply say `properties` instead of `allowedProperties` because we intend to use this format as a general purpose format, and it may end up being used in other places to semantically mean `bannedProperties`.

@JoshuaKGoldberg thumbs up from you?

**Comment by JoshuaKGoldberg** [comment]
![Kramer from Seinfeld giving a thumbs up and leaving the apartment](https://github.com/user-attachments/assets/8fff892f-9ae3-463d-a3ee-2d7639afef45)

**Comment by Harpush** [comment]
@kirkwaiblinger @JoshuaKGoldberg I tried looking at how to implement it as the `accepting prs` label was added.
I can't see a way currently to do it automatically. It seems each place using the `typeMatchesSomeSpecifier` function will need to adjust to the case of a property. Any other ideas?
For example the `no-deprecated` rule is currently using `Identifier` node which needs to check parent whether it is under a member expression.

A somewhat "wild" idea is to pass a node (and not a type) to the `typeMatchesSomeSpecifier` and check the AST upwards.
For example given the type:
```ts
class Test {
  /** @deprecated */
  prop: string;
}

new Test().prop
```  
The rule visitor will stop on the `prop` identifier. Getting the type `string` from it which is a dead-end.
The `allow` options for this will be `{ name: 'Test', properties: ['prop'] }` so based on what I said we could pass the `Identifier` node and check the parent if we have a `MemberExpression` and then check if parent type matches the specifier name.

This is also not a full proof and requires some changes in some places using the `typeMatchesSomeSpecifier`

**Comment by JoshuaKGoldberg** [comment]
> It seems each place using the typeMatchesSomeSpecifier function will need to adjust to the case of a property. 

Yeah, I think that's what needs to happen. Which is a little unfortunate. But I can't think of a better way to do it. If you can spot a way to refactor the code to be less duplicate-y, that would be nice! 😄 

> A somewhat "wild" idea is to pass a node (and not a type) ...

🤔 interesting. I was thinking adding an optional `property?: string` (or, `propertyName`?) to `typeMatchesSpecifier` would be enough. Hard to say without a proof of concept or two to look at though.

I have to admit, this area of code isn't one we work in a lot - so it's a little hard to give definitive advice on it.

**Comment by Harpush** [comment]
property parameter was my initial thought too but that will lead to the duplication.
Essentially what we want in simple terms is to check what's the current position is in the nodes tree. Once we know that we can check whether we are on a property and whether the type is correct.
The reason for node and not type is only nodes has parent access.
Not sure this will work fully though... Might try to create a poc

**Comment by alcuadrado** [comment]
I may have some time in the next week or so to send a PR fixing this. Can I somehow help? Or are you working on it, @Harpush?

To me being able to enable this rule when using `node:test` is critical, and I can't do it now because of `it.skip`, `describe.skip` and the like, so I'm willing to help fix it.

**Comment by Harpush** [comment]
@alcuadrado not currently... Tried a bit but I don't have much time.

**Comment by alcuadrado** [comment]
I'm trying this approach, where I added a `properties` field, but I'm a bit blocked, given that I have to navigate the tree/types bottom-up. The reason is that the `TypeOrValueSpecifier` will be called with, for example, the `skip` of `it.skip`, so I match the property first, and then try to match where it was declared.

Here's my current attempt.

```diff
diff --git a/packages/type-utils/src/TypeOrValueSpecifier.ts b/packages/type-utils/src/TypeOrValueSpecifier.ts
index 4c8f1211a..4d077d8a6 100644
--- a/packages/type-utils/src/TypeOrValueSpecifier.ts
+++ b/packages/type-utils/src/TypeOrValueSpecifier.ts
@@ -24,6 +24,11 @@ export interface FileSpecifier {
    * A specific file the types or values must be declared in.
    */
   path?: string;
+
+  /**
+   * Property name(s) to match on the matched type.
+   */
+  properties?: string | string[];
 }
 
 /**
@@ -37,6 +42,11 @@ export interface LibSpecifier {
    * Type or value name(s) to match on.
    */
   name: string | string[];
+
+  /**
+   * Property name(s) to match on the matched type.
+   */
+  properties?: string | string[];
 }
 
 /**
@@ -55,6 +65,11 @@ export interface PackageSpecifier {
    * Package name the type or value must be declared in.
    */
   package: string;
+
+  /**
+   * Property name(s) to match on the matched type.
+   */
+  properties?: string | string[];
 }
 
 /**
@@ -98,6 +113,21 @@ export const typeOrValueSpecifiersSchema = {
           path: {
             type: 'string',
           },
+          properties: {
+            oneOf: [
+              {
+                type: 'string',
+              },
+              {
+                items: {
+                  type: 'string',
+                },
+                minItems: 1,
+                type: 'array',
+                uniqueItems: true,
+              },
+            ],
+          },
         },
         required: ['from', 'name'],
         type: 'object',
@@ -124,6 +154,21 @@ export const typeOrValueSpecifiersSchema = {
               },
             ],
           },
+          properties: {
+            oneOf: [
+              {
+                type: 'string',
+              },
+              {
+                items: {
+                  type: 'string',
+                },
+                minItems: 1,
+                type: 'array',
+                uniqueItems: true,
+              },
+            ],
+          },
         },
         required: ['from', 'name'],
         type: 'object',
@@ -153,6 +198,21 @@ export const typeOrValueSpecifiersSchema = {
           package: {
             type: 'string',
           },
+          properties: {
+            oneOf: [
+              {
+                type: 'string',
+              },
+              {
+                items: {
+                  type: 'string',
+                },
+                minItems: 1,
+                type: 'array',
+                uniqueItems: true,
+              },
+            ],
+          },
         },
         required: ['from', 'name', 'package'],
         type: 'object',
@@ -174,7 +234,29 @@ export function typeMatchesSpecifier(
     if (typeof specifier === 'string') {
       return specifierNameMatches(type, specifier);
     }
-    if (!specifierNameMatches(type, specifier.name)) {
+    if (specifier.properties) {
+      if (!specifierNameMatches(type, specifier.properties)) {
+        return false;
+      }
+
+      const symbol = type.getSymbol();
+      if (!symbol) {
+        return false;
+      }
+
+      const anyParentMatches =
+        symbol.getDeclarations()?.some(declaration => {
+          const parentType = program
+            .getTypeChecker()
+            .getTypeAtLocation(declaration.parent);
+
+          return specifierNameMatches(parentType, specifier.name);
+        }) ?? false;
+
+      if (!anyParentMatches) {
+        return false;
+      }
+    } else if (!specifierNameMatches(type, specifier.name)) {
       return false;
     }
     const symbol = type.getSymbol() ?? type.aliasSymbol;
```

I can't find a way to get from a `declaration.parent` to a `type` that doesn't error.

**Comment by Harpush** [comment]
I think parent only exists for the AST... You can't get a parent from types - hence complexity

**Comment by alcuadrado** [comment]
Would something like this work better?


```diff
diff --git a/packages/eslint-plugin/src/rules/no-floating-promises.ts b/packages/eslint-plugin/src/rules/no-floating-promises.ts
index 50965093e..664bea2f3 100644
--- a/packages/eslint-plugin/src/rules/no-floating-promises.ts
+++ b/packages/eslint-plugin/src/rules/no-floating-promises.ts
@@ -241,11 +241,19 @@ export default createRule<Options, MessageId>({
       }
 
       const type = services.getTypeAtLocation(node.callee);
+      let memberExpressionTypes: {object: ts.Type; property: ts.Type;} | undefined;
+      if (node.callee.type === AST_NODE_TYPES.MemberExpression) {
+        const object = services.getTypeAtLocation(node.callee.object);
+        const property = services.getTypeAtLocation(node.callee.property);
+
+        memberExpressionTypes = {object, property};
+      }
 
       return typeMatchesSomeSpecifier(
         type,
         allowForKnownSafeCalls,
         services.program,
+        memberExpressionTypes,
       );
     }
```

It would have to be applied to all the rules that use `typeMatchesSomeSpecifier`, which are:
* `no-deprecated`
* `only-throw-error`
* `no-floating-promises`
* `restrict-template-expressions`
* `no-misused-spread`

**Comment by alcuadrado** [comment]
I'm afraid I won't have enough time to tackle this. It's more complex than I expected. I'm using this in the meanting

```
      "@typescript-eslint/no-floating-promises": [
        "error",
        {
          // Node <24 uses floating promises in tests and suites. It was removed in Node 24,
          // but at the time of writing they are considering reverting it, so we allow them
          // here. See: https://github.com/nodejs/node/pull/58282
          allowForKnownSafeCalls: [
            {
              from: "package",
              name: ["describe", "suite", "it", "test"],
              package: "node:test",
            },
            // These are more permissive than needed, as they allow any call to functions
            // with these names to float a promise. They can be removed if Node 24+
            // doesn't revert the removal of floating promises, or if this gets resolved
            // https://github.com/typescript-eslint/typescript-eslint/issues/10740
            "skip",
            "todo",
            "only",
          ],
        },
      ],
```
